### PR TITLE
Add dwb tests

### PR DESCRIPTION
**Pull request by wilcobonestroo**
## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Primary OS tested on | (Ubuntu in VM) |

---

## Description of contribution in a few bullet points

* Added tests for nav_2d_utils in dwb_controller.

## Future work that may be required in bullet points
The test `AdjustResolutionSimple` in `path_ops_test.cpp` should be completed. What is the expected behavior of `adjustPlanResolution`? I expected that the gaps between points would not be bigger than the resolution.

### CODE REVIEW

File: nav2_dwb_controller/nav_2d_utils/test/2d_utils_test.cpp [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  std::vector<geometry_msgs::msg::PoseStamped> poses;
+  nav_msgs::msg::Path path = posesToPath(poses);
+
+  EXPECT_EQ(path.poses.size(), 0ul);
+}
+
+
+TEST(nav_2d_utils, PosesToPathNonEmpty)
+{
+  std::vector<geometry_msgs::msg::PoseStamped> poses;
```
Comment (SteveMacenski) [inline nav2_dwb_controller/nav_2d_utils/test/2d_utils_test.cpp:56]: Populate these with unique values and then iterate over them and make sure they're correct.

Comment (SteveMacenski) [reply nav2_dwb_controller/nav_2d_utils/test/2d_utils_test.cpp:56]: + header / stamp too

Comment (SteveMacenski) [reply nav2_dwb_controller/nav_2d_utils/test/2d_utils_test.cpp:56]: Apply this comment to all below. We need to actually test the code as working, not just exercise it. Look at the functions and what they're doing and then design tests to make sure that everything on the output is as it should be. So properly populate the inputs fully and then check the outputs fully.

File: nav2_dwb_controller/nav_2d_utils/test/path_ops_test.cpp [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  nav_2d_msgs::msg::Path2D out = adjustPlanResolution(in, RESOLUTION);
+  EXPECT_EQ(out.poses.size() > 2, true);
+
+  for (unsigned int i = 1; i < out.poses.size(); i++) {
+    pose1 = out.poses[i - 1];
+    pose2 = out.poses[i];
+
+    double sq_dist = (pose1.x - pose2.x) * (pose1.x - pose2.x) +
+      (pose1.y - pose2.y) * (pose1.y - pose2.y);
+    std::cout << "Distance is " << sqrt(sq_dist) << std::endl;
```
Comment (SteveMacenski) [inline nav2_dwb_controller/nav_2d_utils/test/path_ops_test.cpp:73]: Don't print things when you should check them for content with an `EXPECT_XYZ` test.

This should be less than `resolution * resolution * 4.0`, as the code suggests

**Review by SteveMacenski — CHANGES_REQUESTED** [review]

File: nav2_dwb_controller/nav_2d_utils/test/2d_utils_test.cpp (hunk 1 above)
Comment (wilcobonestroo) [reply nav2_dwb_controller/nav_2d_utils/test/2d_utils_test.cpp:56]: I can fill the position data for a `PoseStamped` with lines such as `pose1.pose.position.x = 1.0;`. However, how do I fill the header manually? E.g. `pose2.header.seq = 2;` says that header has no member named `seq`. So, how to populate a header for testing? Is is only about the `stamp`?

File: nav2_dwb_controller/nav_2d_utils/test/path_ops_test.cpp (hunk 2 above)
Comment (wilcobonestroo) [reply nav2_dwb_controller/nav_2d_utils/test/path_ops_test.cpp:73]: The documentation in the header file says: `@param resolution desired distance between waypoints` I expected that a resolution of 20 (m) would result in a path where the distance between points would be 20. I saw the comment in the code, but I do not see why `resolution * resolution * 4.0` is used.

Comment (wilcobonestroo) [reply nav2_dwb_controller/nav_2d_utils/test/path_ops_test.cpp]: E.g. in the test I have used a path of 100m with a resolution of 20m. It makes a path with 3 points and a distance of 50m between them. I had expected 6 points with 20m in between.

Comment (wilcobonestroo) [reply nav2_dwb_controller/nav_2d_utils/test/path_ops_test.cpp]: I think I get it now. You want to have one point in every 4 "pixels". So, if the resolution is 20, you use 2 * 20 squared as the area where at least one point should be. If `min_sq_resolution` is renamed to `max_sq_resolution` it makes more sense to me.

File: nav2_dwb_controller/nav_2d_utils/src/path_ops.cpp [hunk 3]
```diff
@@ -52,14 +52,14 @@ nav_2d_msgs::msg::Path2D adjustPlanResolution(
   global_plan_out.poses.push_back(last);
 
   // we can take "holes" in the plan smaller than 2 grid cells (squared = 4)
-  double min_sq_resolution = resolution * resolution * 4.0;
+  double max_sq_resolution = resolution * resolution * 4.0;
```
Comment (SteveMacenski) [inline nav2_dwb_controller/nav_2d_utils/src/path_ops.cpp:55]: Please change this variable back - its not good form to change code in while developing tests unless you're fixing a legitimate bug. I also think as it was made more sense.

File: nav2_dwb_controller/nav_2d_utils/test/path_ops_test.cpp [hunk 4]
```diff
@@ -62,17 +62,19 @@ TEST(path_ops_test, AdjustResolutionSimple)
...
     pose2 = out.poses[i];
 
     double sq_dist = (pose1.x - pose2.x) * (pose1.x - pose2.x) +
       (pose1.y - pose2.y) * (pose1.y - pose2.y);
-    // TODO(wilcobonestroo): remove next line
-    std::cout << "Distance is " << sqrt(sq_dist) << " max: " <<
-      sqrt(RESOLUTION * RESOLUTION * 4.0) << std::endl;
-    EXPECT_EQ(sqrt(sq_dist) <= sqrt(RESOLUTION * RESOLUTION * 4.0), true);
+
+    EXPECT_EQ(sqrt(sq_dist) <= max_length, true);
```
Comment (SteveMacenski) [inline nav2_dwb_controller/nav_2d_utils/test/path_ops_test.cpp:77]: You can use `EXPECT_TRUE` instead here

File: nav2_dwb_controller/nav_2d_utils/test/path_ops_test.cpp [hunk 5]
```diff
@@ -62,17 +62,19 @@ TEST(path_ops_test, AdjustResolutionSimple)
...
   }
+  std::cout << "number of points " << out.poses.size();
```
Comment (SteveMacenski) [inline nav2_dwb_controller/nav_2d_utils/test/path_ops_test.cpp:79]: remove the cout

**Review by SteveMacenski — CHANGES_REQUESTED** [review]

File: nav2_dwb_controller/nav_2d_utils/test/2d_utils_test.cpp (hunk 1 above)
Comment (SteveMacenski) [reply nav2_dwb_controller/nav_2d_utils/test/2d_utils_test.cpp:57]: `seq` doesn't exist in ROS2, its only `stamp` and `frame_id`, so I'd fill in with some frame string and a random stamp.

File: nav2_dwb_controller/nav_2d_utils/test/path_ops_test.cpp (hunk 2 above)
Comment (SteveMacenski) [reply nav2_dwb_controller/nav_2d_utils/test/path_ops_test.cpp]: res*res*4 is the same as (2 * resolution)^2, which makes sense when looking for distances in grid-world.

Comment (SteveMacenski) [reply nav2_dwb_controller/nav_2d_utils/test/path_ops_test.cpp]: This function seems pretty straight forward:
- Given a path of variable resolution and a desired static resolution:
- Compute the minimum distance between points to act upon (that 4 * res * res) 
- if pt `i` and pt `i-1` is has more distance between them and then minimum distance, we scale the coordinate locations along the line segment made up of them
- Repeat

Comment (wilcobonestroo) [reply nav2_dwb_controller/nav_2d_utils/test/path_ops_test.cpp]: In that case I would still call it maximum distance.

File: nav2_dwb_controller/nav_2d_utils/test/2d_utils_test.cpp [hunk 6]
```diff
@@ -0,0 +1,162 @@
+/*
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2018, Locus Robotics
```
Comment (SteveMacenski) [inline nav2_dwb_controller/nav_2d_utils/test/2d_utils_test.cpp:4]: Are you locus robotics? :-)

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
LGTM, but update the copyrights, lets see what the new coverage report says!

File: nav2_dwb_controller/nav_2d_utils/test/2d_utils_test.cpp (hunk 6 above)
Comment (wilcobonestroo) [reply nav2_dwb_controller/nav_2d_utils/test/2d_utils_test.cpp]: Sorry. I had just copy pasted that part from another file. I have corrected this.

**Review by SteveMacenski — APPROVED** [review]

**Comment by SteveMacenski** [comment]
Waiting in CI to generate coverage report

**Pull request merged**